src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc

```diff
@@ -0,0 +1,294 @@
+  auto self = static_cast<CFStreamEndpoint*>(client_callback_info);
+
+  switch (type) {
+    case kCFStreamEventOpenCompleted:
+      self->open_event_.SetReady();
+      break;
+    case kCFStreamEventCanAcceptBytes:
+    case kCFStreamEventEndEncountered:
+      self->write_event_.SetReady();
+      break;
+    case kCFStreamEventErrorOccurred: {
+      auto status = CFErrorToStatus(CFWriteStreamCopyError(stream));
+      gpr_log(GPR_ERROR, "CFStream Write error: %s", status.ToString().c_str());
+      self->open_event_.SetShutdown(status);
+      self->read_event_.SetShutdown(status);
+      self->write_event_.SetShutdown(status);
+    } break;
+    default:
+      GPR_UNREACHABLE_CODE(return);
+  }
+}
+
+CFStreamEndpoint::CFStreamEndpoint(std::shared_ptr<CFEventEngine> engine,
+                                   MemoryAllocator memory_allocator)
+    : engine_(std::move(engine)),
+      memory_allocator_(std::move(memory_allocator)),
+      open_event_(engine_.get()),
+      read_event_(engine_.get()),
+      write_event_(engine_.get()) {
+  open_event_.InitEvent();
+  read_event_.InitEvent();
+  write_event_.InitEvent();
+}
+
+CFStreamEndpoint::~CFStreamEndpoint() {
+  CFReadStreamClose(cf_read_stream_);
+  CFWriteStreamClose(cf_write_stream_);
+
+  open_event_.SetShutdown(absl::OkStatus());
+  read_event_.SetShutdown(absl::OkStatus());
+  write_event_.SetShutdown(absl::OkStatus());
+  open_event_.DestroyEvent();
+  read_event_.DestroyEvent();
+  write_event_.DestroyEvent();
+}
+
+void CFStreamEndpoint::Read(absl::AnyInvocable<void(absl::Status)> on_read,
+                            SliceBuffer* buffer, const ReadArgs* /* args */) {
+  read_event_.NotifyOn(new PosixEngineClosure(
+      [this, on_read = std::move(on_read),
+       buffer](absl::Status status) mutable {
+        if (status.ok()) {
+          DoRead(std::move(on_read), buffer);
+        } else {
+          on_read(status);
+        }
+      },
+      false /* is_permanent*/));
+}
+
+void CFStreamEndpoint::DoRead(absl::AnyInvocable<void(absl::Status)> on_read,
+                              SliceBuffer* buffer) {
+  int buffer_size = 8192;
+  auto buffer_index =
+      buffer->AppendIndexed(Slice(memory_allocator_.MakeSlice(buffer_size)));
+
+  CFIndex read_size = CFReadStreamRead(
+      cf_read_stream_,
+      internal::SliceCast<MutableSlice>(buffer->MutableSliceAt(buffer_index))
+          .begin(),
+      buffer_size);
+
+  if (read_size < 0) {
+    auto status = CFErrorToStatus(CFReadStreamCopyError(cf_read_stream_));
+    gpr_log(GPR_ERROR, "CFStream read error: %s", status.ToString().c_str());
+    on_read(status);
+    return;
+  }
+
+  buffer->RemoveLastNBytes(buffer->Length() - read_size);
+  on_read(absl::OkStatus());
+}
+
+void CFStreamEndpoint::Write(absl::AnyInvocable<void(absl::Status)> on_writable,
+                             SliceBuffer* data, const WriteArgs* /* args */) {
+  write_event_.NotifyOn(new PosixEngineClosure(
```

Comment:
I'm a bit lost here, what is actually writing the data? I believe the lockfree event here is triggered when a write finishes, and no write has been initiated yet. Does this work?